src/python/pants/engine/internals/native.py

```diff
@@ -34,35 +50,85 @@
 logger = logging.getLogger(__name__)
 
 
+class IntrinsicError(ValueError):
+    """Exceptions raised for failures within intrinsic methods implemented in Rust."""
+
+
+class IncorrectProductError(TypeError):
+    """Exceptions raised when a rule's return value doesn't match its declared type."""
+
+
 class Externs:
     """Methods exposed from Python to Rust.
 
     TODO: These could be implemented in Rust in `externs.rs` via the cpython API.
     """
 
-    def __init__(self, lib):
+    def __init__(self, lib: ModuleType):
         self.lib = lib
 
+    # NB: This flag is exposed for testing error handling in CFFI methods. This should never be set
+    # in normal pants usage.
     _do_raise_keyboardinterrupt = bool(os.environ.get("_RAISE_KEYBOARDINTERRUPT_IN_EXTERNS", False))
 
-    def is_union(self, input_type):
+    def is_union(self, input_type: Type) -> bool:
         """Return whether or not a type is a member of a union."""
         # NB: This check is exposed for testing error handling in CFFI methods. This code path should
         # never be active in normal pants usage.
         return union.is_instance(input_type)
 
-    def create_exception(self, msg):
-        """Given a utf8 message string, create an Exception object."""
-        return Exception(msg)
+    def create_value_error(self, msg: str) -> IntrinsicError:
+        """Given a utf8 message string, create an exception signalling a value error."""
+        return IntrinsicError(msg)
+
+    def create_type_error(self, msg: str) -> IncorrectProductError:
+        """Given a utf8 message string, create an exception signalling a type error."""
+        return IncorrectProductError(msg)
+
+    # A specification for how the native engine interacts with @rule coroutines:
+    # - coroutines may await on a `Get` or a tuple of `Get`.
+    # - we will send back a single `Any` or a tuple of `Any`, depending upon the variant of `Get`.
+    # - a coroutine will eventually return a single `Any`.
+    RuleSend = Union[Any, Tuple[Any, ...]]
+    RuleYield = Union[Get, Tuple[Get, ...]]
+    RuleCoroutine = Coroutine[RuleSend, RuleYield, Any]
+
+    @overload
+    def _send_to_coroutine(self, func: RuleCoroutine, arg: RuleSend) -> RuleYield:
```

Comment:
I'm not sure how these overloads of `_send_to_coroutine` are different from just having one signature with `arg: Union[RuleSend, Exception]`

Reply:
I don't think they are! I will remove them.